hip/multigrid/amgx_pgm_kernels.hip.cpp

```diff
@@ -0,0 +1,131 @@
+Copyright (c) 2017-2022, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/multigrid/amgx_pgm_kernels.hpp"
+
+
+#include <memory>
+
+
+#include <thrust/device_ptr.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/reduce.h>
+#include <thrust/sort.h>
+#include <thrust/tuple.h>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/multigrid/amgx_pgm.hpp>
+
+
+#include "hip/base/types.hip.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace hip {
+/**
+ * @brief The AMGX_PGM solver namespace.
+ *
+ * @ingroup amgx_pgm
+ */
+namespace amgx_pgm {
+
+
+template <typename IndexType>
+void sort_agg(std::shared_ptr<const DefaultExecutor> exec, IndexType num,
+              IndexType* row_idxs, IndexType* col_idxs)
+{
+    auto it = thrust::make_zip_iterator(
+        thrust::make_tuple(thrust::device_pointer_cast(row_idxs),
+                           thrust::device_pointer_cast(col_idxs)));
+    thrust::sort(thrust::device, it, it + num);
+}
+
+GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_AMGX_PGM_SORT_AGG_KERNEL);
+
+
+template <typename ValueType, typename IndexType>
+void sort_row_major(std::shared_ptr<const DefaultExecutor> exec, size_type nnz,
+                    IndexType* row_idxs, IndexType* col_idxs, ValueType* vals)
+{
+    // workaround for CUDA 9.2 Thrust: Their complex<> implementation is broken
```

Comment:
Is this comment relevant for HIP thrust as well ?

Reply:
The comment was probably copied from my code: Yes, if you compile HIP for NVIDIA, you use NVIDIA Thrust, and rocThrust was forked from NVIDIA Thrust, so it might be that this issue is also present in an older rocThrust version.